django/db/models/query_utils.py

```diff
@@ -78,6 +79,26 @@ def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize
         query.promote_joins(joins)
         return clause
 
+    def check(self, against, using=DEFAULT_DB_ALIAS):
+        """
+        Do a database query to check if the expressions of the Q instance
+        matches against the `against` expressions.
+        """
+        # Avoid circular imports
+        from django.db.models.sql import Query
+        from django.db.models.sql.constants import SINGLE
+        from django.db.models import Value, ExpressionWrapper, BooleanField
+
+        query = Query(None)
+        for name, value in against.items():
+            if not hasattr(value, 'resolve_expression'):
+                value = Value(value)
+            query.add_annotation(value, name, select=False)
+        # This will raise a FieldError if a needed field is missing from against
+        query.add_annotation(ExpressionWrapper(self, BooleanField()), '_check')
+        compiler = query.get_compiler(using=using)
+        return bool(compiler.execute_sql(SINGLE)[0])
+
```

Comment:
The discussion about making this method public makes me wonder if we should not make it a private method of `BaseConstraint` for the time being.

Reply:
@charettes I'd imagined this being user-accessible (public). Originally, I'd imagined folks implementing this in Python (`return x > 18` sort of thing). Your suggestion to go via the DB is significantly better (more robust) but I'd still imagine folks wanting to implement constraints for which a simple Python check would be enough. Documenting the hook would allow that... 🤔 (Discuss :)

Reply:
> I'd still imagine folks wanting to implement constraints for which a simple Python check would be enough. 

@carltongibson in this case we need to document `Constraint.validate` and not necessarily `Q.check` don't we?

Reply:
Yes, that covers my use-case. (Sorry)

Reply:
No worries, just wanted to make sure we were on the same line.